src/handwritten/_lib_time.py

```diff
@@ -29,21 +30,30 @@ class AbsoluteTime(ctypes.Structure):
     MAX_FS = 10**9
     MAX_YS = 10**9
 
+    _EPOCH_1904 = ht_datetime(1904, 1, 1, tzinfo=timezone.utc)
+
+    def _convert_to_desired_timezone(self, expected_time_utc: Union[std_datetime, ht_datetime], tzinfo: Optional[timezone] = None):
+        current_time_utc = ht_datetime.now(timezone.utc)
+        desired_timezone_offset = current_time_utc.astimezone(tz=tzinfo).utcoffset()
+        desired_expected_time = expected_time_utc + desired_timezone_offset
+        new_datetime = ht_datetime(
+            desired_expected_time.year,
+            desired_expected_time.month,
+            desired_expected_time.day,
+            desired_expected_time.hour,
+            desired_expected_time.minute,
+            desired_expected_time.second,
+            desired_expected_time.microsecond,
+            desired_expected_time.femtosecond,
+            desired_expected_time.yoctosecond,
+            tzinfo = tzinfo)
+        return new_datetime
```

Comment:
This looks like a bespoke implementation of [datetime.astimezone](https://docs.python.org/3/library/datetime.html#datetime.datetime.astimezone) (hightime has the same method). That is dangerous...

Reply:
Ahhh, ok. Add a comment to that effect :)

Reply:
More questions about this timezone conversion...

Does the timezone offset account for daylight saving time? If so, the timezone offset now is not necessarily the timezone offset for some past or future date. For example, the offset would change by 1 hour on March 10th in the US.

Is this timezone conversion even used?
```
PS D:\dev\nidaqmx-python> rg "convert_timestamp_to_time|to_datetime" .\generated\nidaqmx\
.\generated\nidaqmx\_grpc_time.py
38:def convert_timestamp_to_time(ts: GrpcTimestamp, tzinfo: Optional[timezone] = None) -> ht_datetime:

.\generated\nidaqmx\_grpc_interpreter.py
21:from nidaqmx._grpc_time import convert_time_to_timestamp, convert_timestamp_to_time
2234:        return convert_timestamp_to_time(response.value)
3339:        return convert_timestamp_to_time(response.timestamp)

.\generated\nidaqmx\_lib_time.py
61:    def to_datetime(self, tzinfo: Optional[timezone] = None) -> ht_datetime:

.\generated\nidaqmx\_library_interpreter.py
3819:        return value.to_datetime()
5845:        return timestamp.to_datetime()
```

In the C API, I think the time trigger attributes like `DAQmxSetSyncPulseTimeWhen` pass the timestamp through without doing any timezone conversions. However, the channel calibration attributes like `DAQmxGetAIChanCalCalDate` convert to/from local time. They also break everything out into year, month, day, hour, minutes, so they would not use this code.

Reply:
> Does the timezone offset account for daylight saving time?

Very good question, let me look into this a bit further before getting back to you.

> Is this timezone conversion even used?

Yes it is used, in python's `to_datetime` and grpc's `convert_timestamp_to_time`. You may refer to my reply in the [WaitForValidTimestamp PR](https://github.com/ni/nidaqmx-python/pull/510#discussion_r1498963296) to check that it's used. 

>In the C API, I think the time trigger attributes like DAQmxSetSyncPulseTimeWhen pass the timestamp through without doing any timezone conversions. 

If what they take is a timestamp value, then I believe the timestamp value in itself should already have taken into consideration the timezone conversion.

> the channel calibration attributes like DAQmxGetAIChanCalCalDate convert to/from local time. They also break everything out into year, month, day, hour, minutes, so they would not use this code. Also, I suspect that channel calibration dates may not return values before 1970.

It seems like we currently don't have support for these attributes in python, would this also be part of the things to port over in daqmx-python? cc @WayneDroid

Reply:
> > Does the timezone offset account for daylight saving time?
> 
> Very good question, let me look into this a bit further before getting back to you.

One way to look at this is to write a unit test that compares the result of the standard `.astimezone()` method with your function for various dates in the current year (such as the first of each month, which should include some DST and some non-DST dates in regions that have DST).

> > Is this timezone conversion even used?
>
> Yes it is used, in python's to_datetime and grpc's convert_timestamp_to_time. You may refer to my reply in the https://github.com/ni/nidaqmx-python/pull/510#discussion_r1498963296 to check that it's used.

Thanks. The thing I missed is that the default arg `tzinfo=None` means "convert to local time", not "don't convert timezone".

> > the channel calibration attributes like DAQmxGetAIChanCalCalDate convert to/from local time. They also break everything out into year, month, day, hour, minutes, so they would not use this code. Also, I suspect that channel calibration dates may not return values before 1970.
>
> It seems like we currently don't have support for these attributes in python, would this also be part of the things to port over in daqmx-python? cc @WayneDroid

In AB#1645316, @zhindes put channel calibration in the "I wouldn't add, but might be useful to add someday" category.

Reply:
@DeborahOoi96 @bkeryan Yup, Channel Calibration properties are not planned to be included in 1.0 release.

Reply:
@bkeryan Unfortunately, the timezone offset did not account for daylight saving time. :( I have modified it now to accommodate daylight savings times but this involves adding a new dependency, `pytz` in the unit tests. This seems to be breaking PR build due to poetry.lock not being updated,  but I am facing this certification error when trying to run poetry lock. 
![image](https://github.com/ni/nidaqmx-python/assets/74756143/fa6629d3-ec85-4842-88b5-6742af646a94)

May I know how to update the dependency in daqmx-python?

As for what went wrong in my previous time conversion function:
1. It used the *local time offset* to calculate the expected datetime to be returned to the user. Which means if I were to run the function in a DST-supported region **at a time in which DST is supported**, it should theoretically work (since astimezone does account for DST).
2. But for the sake of unit test, we want to check if the offset for a future/past date is the same as the offset from a present date, even if they are in the same DST region. This would vary in the cases of DST and be inconsistent.

So I used pytz.localize to help me with this. It will directly convert the `expected_time_utc` to the expected timezone while considering DST.

Reply:
But adding a new dependency, `pytz` in the unit tests seems to be breaking PR build due to poetry.lock not being updated,  but I am facing this certification error when trying to run poetry lock. 
![image](https://github.com/ni/nidaqmx-python/assets/74756143/fa6629d3-ec85-4842-88b5-6742af646a94)

May I know how to update the dependency in daqmx-python?